### DISCUSSION TITLE
Documentation for the API to change LLMs

### DISCUSSION THREAD

**Discussion by wlauppe** [karthink/gptel#401] [General]
Hello Karthik, you asked in the issues, from a user perspective, what a good gptel-ui would look like. 
(https://github.com/karthink/gptel/issues/249 and related issues)

I thought about this for a bit. One interesting thing i miss, is a very simple "Quickchange Menu"  to change the LLM-Model. 

Here is a first draft. 
![screenshot_20240928_074718](https://github.com/user-attachments/assets/87d117a1-28d8-449f-904f-12b6ffe29d85)
```
#+begin_src elisp
(require 'transient)

(defvar my-model-variable nil "Variable to store the selected model.")

(defun set-my-model-variable (value)
  "Set the my-model-variable to the given VALUE."
  (setq my-model-variable value))

(transient-define-prefix my-select-model-menu ()
  "Transient for selecting AI models."
  ["Select AI Model"
   ("1" "claude opus"   (lambda () (interactive) (set-my-model-variable "claude opus")))
   ("2" "gpt4o"         (lambda () (interactive) (set-my-model-variable "gpt4o")))
   ("3" "claude sonnet" (lambda () (interactive) (set-my-model-variable "claude sonnet")))
   ("4" "gpt4-turbo"    (lambda () (interactive) (set-my-model-variable "gpt4-turbo")))
   ])

(global-set-key (kbd "C-c m") 'my-select-model-menu)
#+end_src
```
If you put this code in an orgmode file, and press ```C-c  C-c```  and  ```C-c m``` the code works, but it is not connected to any function. 

What need is info about, how to wire this code to gptel. Of course it should not only set the string to but actually change the model. 

is there any existing, documentation on how to do that, or could you provide some help?

I tried to figure it out on my own by looking at 
```transient-define-prefix gptel-menu ()``` in ```gptel-transient.el``` but this code is way above my novice understanding of lisp. 

Can you help? Thanks a lot!

**Comment by karthink** [comment]
See #189, specifically the code block under "Helper functions".  Let me know if you need more help.

**Reply by wlauppe** [reply]
thanks for your quick reply and pointing me into the right direction. 
 >Let me know if you need more help.

Actually I do:

for example I execute ```gptel-menu``` and see this:
![screenshot_20240930_095754](https://github.com/user-attachments/assets/9fb214fa-ebf0-4b2f-accc-8129b60874e3)
Then i change the model via the menu to another model. 
Now i want to change it via back to the original setting via api calls. 
From #189  I pieced together, that I must set the variables as follows: 

```
 (setq gptel-model "claude-3-opus-20240229")
 (setq gptel-backend (cdr (assoc "Claude" gptel--known-backends))))
```
if I execute this, the backend is changed but the model is not the correct one. 
Executing ```gptel-menu```
shows the following: 
![screenshot_20240930_095621](https://github.com/user-attachments/assets/0003438a-c8dc-48d2-9370-c34a85b330f5)
what am I missing?

**Reply by wlauppe** [reply]
additional information, im on your ```feature-capabilities``` branch. Did the syntax change there?
a recent commit of yours says: 
```Change models from strings to symbols```
could you tell for both the feature-capabilities and the master branch on how the two setq lines should look like?

**Reply by karthink** [reply]
> additional information, im on your `feature-capabilities` branch. Did the syntax change there?

Ah yeah, if you're on the image support branch you should be doing the equivalent of

```elisp
 (setq gptel-model 'claude-3-opus-20240229)
 ```

That said, it should just issue a warning, convert the model to a symbol internally and keep working.  Maybe I missed a step somewhere.  You can check if the above works.

**Reply by wlauppe** [reply]
Yes the above works on the ```feature-capabilities``` branch. Thanks a lot 
I don't get a warning if i use 

```emacs-lisp
(setq gptel-model "claude-3-opus-20240229")
 ```
with quotes. 

I have another API question: Is there a hook I can use to get notified every time the model changes? For example, if I want to show the four least recently used models in a menu like the one defined above.
Are the hooks mentioned on the README page the only ones available?

**Reply by karthink** [reply]
> I have another API question: Is there a hook I can use to get notified every time the model changes?

There isn't a `gptel-model-change-hook`.  I'm not sure that it's possible to add one, since there is no single way to change the model.

> if I want to show the four least recently used models in a menu like the one defined above.

You can update the list of recent models from `gptel-pre-response-hook`.

**Comment by karthink** [comment]
Perhaps something like this should be added to the [wiki](https://github.com/karthink/gptel/wiki/Commonly-requested-features).

**Comment by wlauppe** [comment]
für the record, if somebody stumbles over this discussion and wants to implement something similar, thats the working code 

```emacs-lisp
(require 'transient)

(defun set-gptel-model-and-backend (model backend)
  "Set the gptel-model and gptel-backend variables based on the selected MODEL and BACKEND."
  (setq gptel-model model)
  (setq gptel-backend (cdr (assoc backend gptel--known-backends)))
  (message "\"%s:%s\" selected" backend model))

(transient-define-prefix gptel-model-picker ()
  "Transient for selecting AI models."
  ["Select AI Model"
   ("1" "claude opus"   (lambda () (interactive) (set-gptel-model-and-backend 'claude-3-opus-20240229 "Claude")))
   ("2" "gpt4-turbo"    (lambda () (interactive) (set-gptel-model-and-backend 'gpt-4-turbo "ChatGPT")))
   ("3" "gpt4o"         (lambda () (interactive) (set-gptel-model-and-backend 'gpt-4o "ChatGPT")))
   ("4" "claude sonnet" (lambda () (interactive) (set-gptel-model-and-backend 'claude-3-5-sonnet-20240620 "Claude")))
   ])

(global-set-key (kbd "C-c m") 'gptel-model-picker)
 ```
